### ISSUE TITLE
New rule request: value-no-border-none / value-border-zero

### ISSUE THREAD

**Issue by timothyneiljohnson** [stylelint/stylelint#948]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

A new rule to catch uses of `border: none`, and instead use the preferred `border: 0`.

> What CSS is needed to reproduce this issue?

```css
.testClass {
  border: none
}
```

> Which version of stylelint are you using?

`5.0.1`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

PostCSS plugin

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

No

**Comment by evilebottnawi** [comment]
@timothyneiljohnson
Bad rule, `border: 0;` : `border-width: 0;`, `border: none;`: `border-style: none;` (not `0` width). See spec css https://www.w3.org/TR/CSS21/box.html#border-shorthand-properties.
```
border: hidden;    border-style: hidden;
                   border-width: medium;
                   border-color: <the same as 'color' property>

border: none;      border-style: none;
                   border-width: medium;
                   border-color: <the same as 'color' property>

border: 0;         border-style: none;
                   border-width: 0;
                   border-color: <the same as 'color' property>
```

**Comment by timothyneiljohnson** [comment]
@evilebottnawi - Fair point that both are valid. Although there's some merit to claims that `border: none` can be almost-imperceptibly slower than `border: 0` — I think it's more about being able to enforce style standards / uniformity in our projects. Would love to see this rule!

**Comment by evilebottnawi** [comment]
@timothyneiljohnson fair or not fair, but spec always right.

**Comment by timothyneiljohnson** [comment]
@davidtheclark @evilebottnawi - I've created a stylist plugin allowing the user to choose the convention they prefer between 'border: none' and 'border: 0'. Would greatly appreciate any feedback!

**Comment by timothyneiljohnson** [comment]
https://github.com/timothyneiljohnson/stylelint-value-border-zero

**Comment by davidtheclark** [comment]
Thanks. Why not do PRs to get these rules in core?

**Comment by evilebottnawi** [comment]
@davidtheclark i think this rule not for core, i write above about `border: none;` and `border: 0` has difference effect on style.

**Comment by ntwb** [comment]
I'll also add a comment here, having just tested both @timothyneiljohnson stylelint plugins I'd love to see these included as stylelint rules:
• https://github.com/timothyneiljohnson/stylelint-value-border-zero
• https://github.com/timothyneiljohnson/stylelint-value-shorthand

**Comment by timothyneiljohnson** [comment]
Hi @evilebottnawi, they look decievingly similar but there is technically a difference, and it's bitten me before:
https://codepen.io/timothyneiljohnson/pen/zqwwvY

**Comment by evilebottnawi** [comment]
@timothyneiljohnson because we don't use this rule in core, only as plugins.

**Comment by jeddy3** [comment]
The inclusion of this rule seems quite divisive. Is the argument against its inclusion that `border: none` and `border: 0` have a different effects on style and so it doesn't make sense to *always* enforce one or the other, rather you'd use which is most appropriate for the situation in hand?.

As @timothyneiljohnson points out, his use case is avoid dealing with this difference by always enforcing one notation or the other within his codebase. Perhaps there is scope to include this rule in core if we name the rule so that it doesn't favour one notation over the other (like we've done elsewhere with the likes of `color-hex-case` (rather than `color-hex-no-uppercase`)). As well as clearly documenting the use case in the README?

We can call the rule `border-notation`, which accepts the options `"none"` or `"zero"`?

Thoughts?

**Comment by davidtheclark** [comment]
I have no opinions about this, sorry. Happy to let you all arrive at the conclusion :)

**Comment by ntwb** [comment]
> ...the argument against its inclusion that `border: none` and `border: 0` have a different effects on style and so it doesn't make sense to always enforce one or the other 

This is how I read it, both of these are valid so it makes no sense to enforce one of the other.

> Hi @evilebottnawi, they look decievingly similar but there is technically a difference, and it's bitten me before: https://codepen.io/timothyneiljohnson/pen/zqwwvY

@timothyneiljohnson I think you've answered your own question here, "technically there is a difference" but both are valid, and @evilebottnawi noted those subtle technical differences in https://github.com/stylelint/stylelint/issues/948#issuecomment-198569090

**Comment by evilebottnawi** [comment]
I see this rule as plugin, not in core.

**Comment by davidtheclark** [comment]
Ok --- sounds like everybody's in agreement that this should be a plugin ... And it *is* a plugin, because @timothyneiljohnson made it ... Can we close the issue then?

**Comment by timothyneiljohnson** [comment]
@davidtheclark - I'm good with that, too. It's certainly a divisive rule for inclusion in core.

**Comment by pmweeks98** [comment]
Just putting this here - plugin is not needed now
https://github.com/timothyneiljohnson/stylelint-value-border-zero/issues/2
